proposals/runtime_on_invalidated.md

```diff
@@ -0,0 +1,151 @@
+# Proposal: browser.runtime.onInvalidated event
+
+**Summary**
+
+`browser.runtime.onInvalidated`
+
+This event allows extension contexts which stay alive to be notified when the
+extension context get invalidated.
+
+**Document Metadata**
+
+**Author:** carlosjeurissen
+
+**Sponsoring Browser:** Chrome
+
+**Contributors:** bershanskiy
+
+**Created:** 2025-03-26
+
+**Related Issues:** https://github.com/w3c/webextensions/issues/138
+
+## Motivation
+
+### Objective
+
+This event allows extension contexts which stay alive to be notified when the
+extension context get invalidated.
+
+#### Use Cases
+
+When the extension context gets invalidated, this means extension contexts
+which stay alive can no longer use the chrome APIs or communicate with other
+extension contexts.
+
+For extension contexts, knowing when this happens is crucial for moving
+forward. Alternatives can be used in this situation, like `window.open` instead
+of `browser.tabs.create`. If moving forward is no option, this event can be
+used for initiating the cleanup of any side effects caused in the frames
+extensions have been operating in.
+
+### Known Consumers
+
+Any extension with a content script which applies certain side effects to the
+web page.
+
+## Specification
+
+### Schema
+
+New onInvalidated event:
+```json
+{
+  "name": "onInvalidated",
+  "type": "function",
+  "description": "Fired when the extension context get invalidated.",
```

Comment:
```suggestion
  "description": "Fired when the extension context get invalidated while the underlying document remains valid.",
```

Suggested tweak to the description. Otherwise you would also have to fire the event on page unload and browser shutdown.

This description also precludes non-content scripts. If you want to include non-content scripts I'd like to see a good justification. Regular document have the pagehide, beforeunload and unload events already. Background contexts have runtime.onSuspend.